tmt/steps/execute/__init__.py

```diff
@@ -801,28 +803,108 @@ def prepare_tests(self, guest: Guest, logger: tmt.log.Logger) -> list[TestInvoca
 
     def prepare_scripts(self, guest: "tmt.steps.provision.Guest") -> None:
         """
-        Prepare additional scripts for testing
+        Prepare additional scripts locally in a staging directory and push them.
+
+        Creates a temporary local directory structure mirroring the intended
+        guest filesystem layout. Scripts and their aliases (as symbolic links)
+        are placed into this structure. The entire structure is then pushed
+        to the guest's root directory in a single operation.
         """
+        staging_root: Optional[Path] = None
+        rendered_templates: list[Path] = []
+
+        try:
+            # Create a temporary local staging directory
+            staging_root = Path(tempfile.mkdtemp(prefix='tmt-scripts-staging-'))
+            self.debug(f"Created local script staging directory: {staging_root}")
 
-        # Make sure scripts directory exists
-        command = Command("mkdir", "-p", f"{guest.scripts_path}")
+            default_scripts_dest_dir_relative = guest.scripts_path.relative_to('/')
 
-        if not guest.facts.is_superuser:
-            command = Command("sudo") + command
+            for script in self.scripts:
+                if not script.enabled(guest):
+                    continue
+
+                # Determine the final path relative to the guest root '/'
+                if script.destination_path:
+                    # Custom destination path provided
+                    relative_script_path = script.destination_path.relative_to('/')
+                else:
+                    # Default destination path
+                    relative_script_path = (
+                        default_scripts_dest_dir_relative / script.source_filename
+                    )
 
-        guest.execute(command)
+                # Construct the full path within the local staging directory
+                local_script_path = staging_root / relative_script_path
+
+                # Ensure the parent directory exists in the staging area
+                local_script_path.parent.mkdir(parents=True, exist_ok=True)
+
+                # Copy or render the script into the staging area
+                with script as source_path:
+                    if isinstance(script, ScriptTemplate):
+                        # ScriptTemplate renders to a temporary file in __enter__
+                        rendered_path = script._rendered_script_path
+                        assert rendered_path is not None
+                        shutil.copy2(rendered_path, local_script_path)
+                        # Track rendered templates for cleanup, even though __exit__ handles it
+                        rendered_templates.append(rendered_path)
+                    else:
+                        # Regular script, copy the source file
+                        shutil.copy2(source_path, local_script_path)
```

Comment:
How about adding a method to `Script`, something like `render_into(filepath: Path)`? The default implementation would do this `shutil.copy2(SCRIPTS_SRC_DIR / self.source_filename, filepath)` oneliner (`source_path` is the `SCRIPTS_SRC_DIR / self.source_filename`), and implementation in `ScriptTemplate` would use `self._rendered_script_path` instead. Why: the simple task, "store script in a given file", is performed in two ways, depending on which class the `script` is. That screams class-specific implementation is leaking beyond the `Script*` classes.